utils/lib/DataSource.ts

```diff
@@ -0,0 +1,149 @@
+   */
+  getConfigFilePath() {
+    const filePaths = glob.sync(
+      path.join(this.basePath, 'data-sources', this.localPath, '*.+(yml|yaml)')
+    );
+
+    if (!Array.isArray(filePaths) || filePaths.length !== 1) {
+      this.isValid = false;
+      return '';
+    }
+
+    return removeBasePath(filePaths[0], this.basePath);
+  }
+
+  getConfigContent() {
+    return this._getYamlConfigContent();
+  }
+
+  /**
+   * Get the variables for the **Quickstart** mutation.
+   *
+   * @returns The ID for this data source
+   */
+  getMutationVariables(): DataSourceConfig['id'] {
+    return this.config.id;
+  }
+
+  /**
+   * Get the **component-specific** mutation variables.
+   */
+  private _getComponentMutationVariables(
+    dryRun: boolean = true
+  ): DataSourceMutationVariable {
+    const { id, displayName, categoryTerms, keywords, description } =
+      this.config;
+
+    const dataSourceMetadata = {
+      displayName: displayName && displayName.trim(),
+      icon: this._getIconUrl(),
+      install: this._parseInstall(),
+      categoryTerms: categoryTerms && categoryTerms.map((t) => t.trim()),
+      keywords: keywords && keywords.map((k) => k.trim()),
+      description: description && description.trim(),
+    };
+
+    return {
+      dryRun,
+      id,
+      dataSourceMetadata,
+    };
+  }
+
+  public async submitMutation(dryRun = true) {
+    const { data, errors } = await fetchNRGraphqlResults<
+      DataSourceMutationVariable,
+      DataSourceMutationResponse
+    >({
+      queryString: DATA_SOURCE_MUTATION,
+      variables: this._getComponentMutationVariables(dryRun),
+    });
+
+    // filePath may need to be changed for this rework
+    return { data, errors, name: this.localPath };
+  }
+
+  /**
+   * Helper method to get the image URL for the icon.
+   */
+  private _getIconUrl() {
+    const { icon } = this.config;
+    const dirName = path.dirname(this.configPath);
+    const relDirName = removeBasePath(dirName, this.basePath);
+
+    return `${GITHUB_RAW_BASE_URL}/${relDirName}/${icon}`;
+  }
+
+  /**
+   * Helper method to parse the primary and fallback install directives.
+   */
+  private _parseInstall() {
+    const { install } = this.config;
+
+    return {
+      primary: this._parseInstallDirective(install.primary),
+      fallback:
+        install.fallback && this._parseInstallDirective(install.fallback),
+    };
+  }
+
+  /**
+   * Helper method that returns the directive, based on it's type.
+   */
+  private _parseInstallDirective(
+    directive: DataSourceConfigInstallDirective
+  ): DataSourceInstallDirectiveInput {
+    if ('link' in directive) {
+      const { url } = directive.link;
+
+      return {
+        link: {
+          url: url?.trim() ?? '',
+        },
+      };
+    }
+
+    if ('nerdlet' in directive) {
+      const { nerdletId, nerdletState, requiresAccount } = directive.nerdlet;
+
+      return {
+        nerdlet: {
+          nerdletId: nerdletId.trim() ?? '',
```

Comment:
This should throw a `cannot reference trim() of undefined` error if the `nerdletId` is `undefined`. We can use optional chaining to deal with this.
```suggestion
          nerdletId: nerdletId?.trim() ?? '',
```